### PR TITLE
"lr state serialization"

### LINKED ISSUE #2713: Optimizer Learning Rate Policy save state.

### PR DESCRIPTION
**Pull request by dzhwinter**
fix #2713

### CODE REVIEW

File: paddle/optimizer/adadelta_optimizer.cc [hunk 1]
```diff
@@ -27,22 +27,24 @@ void AdadeltaOptimizer::Update(const Tensor* gradient) {
...
+  std::string lr_str = this->lr_policy_->SerializeState(state_len);
+  state.mutable_lr_state()->ParseFromString(lr_str);
 
   TensorToProto(*parameter_, state.mutable_parameter());
   TensorToProto(*accum_gradient_, state.mutable_accum_gradient());
   TensorToProto(*accum_delta_, state.mutable_accum_delta());
   TensorToProto(*update_delta_, state.mutable_update_delta());
   auto str = state.SerializeAsString();
-  *state_len = str.size();
+  *state_len += str.size();
```
Comment (typhoonzero) [inline paddle/optimizer/adadelta_optimizer.cc]: Not sure why need to `+=`?

Comment (dzhwinter) [inline paddle/optimizer/adadelta_optimizer.cc]: because  get the lr_state serialization length here.
```c++
std::string lr_str = this->lr_policy_->SerializeState(state_len)
```
use `=` will cover the lr_state length.

Comment (typhoonzero) [inline paddle/optimizer/adadelta_optimizer.cc]: I see.

File: proto/OptimizerConfig.proto [hunk 2]
```diff
@@ -102,22 +104,17 @@ message AdadeltaOptimizerState {
   optional TensorProto update_delta = 4;
 }
 
+
 message AdagradOptimizerState {
-  // learning rate policy
-  optional double learning_rate = 101;
-  optional double lr_decay_a = 102;
-  optional double lr_decay_b = 103;
+  optional LrPolicyState lr_state = 101;
```
Comment (typhoonzero) [inline proto/OptimizerConfig.proto]: Do not re-use protobuf ids to change the fields. Well, since we it's still under development, this is fine. When people are using this code, we need to consider the backward compatibility.

Comment (dzhwinter) [inline proto/OptimizerConfig.proto]: Thanks for the reminding. That's true when we maintain a project.
But what is the best practice of field number?  Making the field number auto increment by 1 seems not a smart idea. By the way, the field number is only used by `protoc` compiler to represent an entry of protobuf. Even though protobuf support indexed value by field number,  but it should not be used in user's code. Did you ever run into such a code base?

Comment (typhoonzero) [inline proto/OptimizerConfig.proto]: Once the protobuf index is used by messages, it should never change for compatibility, for the old client or somewhere is still using the old `.proto` generated code.

1. Do not change the current index
1. Only append field not modify them.
